nixos/dokuwiki: add support for multi-site, additional plugins and templates

dadada opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

This MR introduces multi-site dokuwiki configurations (e.g. multiple nginx servers) using the <name?> option syntax. It also enables the use of additional plugins and templates similarly to how wordpress.nix does it.

So far everything seems to work fine, but marking as WIP since I'd like to do some manual testing first. I can split this up into two MRs if requested.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

dadada committed:
nixos/dokuwiki: add <name?> option

Enables multi-site configurations.

This break compatibility with prior configurations that expect options
for a single dokuwiki instance in `services.dokuwiki`.

dadada committed:
nixos/dokuwiki: refactor

dadada committed:
nixos/dokuwiki: add plugins and templates options

Adds support for additional plugins and templates similarly to how
wordpress.nix does it.

Plugins and templates need to be packaged as in the example.

dadada commented:
Pinging @1000101 maintainer.

1000101 commented:
@dadada thanks a million for picking up on this, I'll give it a go in a couple of hours.

@aanderse might want to have a look at this, too!

dadada committed:
nixos/dokuwiki: add assertion for usersFile

dadada committed:
nixos/dokuwiki: set default value for usersFile

If usersFile is not set, a file is created along the stateDir that can
hold the users and supports dynamically adding users using the web GUI.

dadada committed:
nixos/dokuwiki: add option disableActions

1000101 commented:
@dadada Let me know when you're happy with the changes and I'll test it.

1000101 commented:
@GrahamcOfBorg eval

dadada committed:
nixos/dokuwiki: dokuwiki user

dadada committed:
nixos/dokuwiki: modify usersFile and aclFile

Use types.str instead of types.path to exclude private information from
the derivation.
Add a warinig about the contents of acl beeing included in the nix
store.

dadada commented:
Thanks for taking an interest. What's your opinion on the way the `users.php` is created? If I understand correctly, this was previously read-only / part of the derivation? Was there some specific reason for this? Should we put a warning if `usersFile` is defined?

dadada commented:
I made two more minor changes. efc9258c97355b4ea6e313c7ab9447d894dbcf34 adds a dokuwiki user that owns all the state directories and files.

07504a1 changes the behaviour of usersFile and aclFile. Since they might contain private information, the contents should not end up in the nix store. If the options are defined, those files are generated at the given paths if they don't already exist.

dadada commented:
> @dadada Let me know when you're happy with the changes and I'll test it.

I'm happy, go ahead.

dadada commented:
I'll need to do a quick reword of the commit messages that change nixos/dokuwiki

dadada commented:
> I'll need to do a quick reword of the commit messages that change nixos/dokuwiki

I'm happy, go ahead. :-)

1000101 commented:
@GrahamcOfBorg test dokuwiki

1000101 commented:
> Thanks for taking an interest. What's your opinion on the way the `users.php` is created? If I understand correctly, this was previously read-only / part of the derivation? Was there some specific reason for this? Should we put a warning if `usersFile` is defined?

You mean users.auth.php (i.e. `usersFile` option)? It was actually not part of the derivation so that  1. credentials would not end up in nix store 2. it would be mutable (users could change their passwords).

Generally, I was trying to have as much options as I can immutable, with obvious ones mutable (https://github.com/NixOS/nixpkgs/pull/77830/files#r368233914).

1000101 commented on nixos/modules/services/web-apps/dokuwiki.nix:
```diff
@@ -266,7 +373,13 @@ in
       "d ${cfg.stateDir}/tmp 0750 ${user} ${group} - -"
-    ];
+      "C ${cfg.usersFile} 0640 ${user} ${group} - ${pkg hostName cfg}/share/dokuwiki/conf/users.auth.php.dist"
+    ] ++ lib.optional (cfg.acl == null) "C ${cfg.aclFile} 0640 ${user} ${group} - ${pkg hostName cfg}/share/dokuwiki/conf/acl.auth.php"
```
```suggestion
    ] ++ lib.optional (cfg.acl == null) "C ${cfg.aclFile} 0640 ${user} ${group} - ${pkg hostName cfg}/share/dokuwiki/conf/acl.auth.php.dist"
```

dadada replied:
See a84c2f4

1000101 replied:
IMO the automatic creation (copying) default user+acl files was nice. Have you found any problems with that?

1000101 commented on nixos/modules/services/web-apps/dokuwiki.nix:
```diff
@@ -18,244 +19,350 @@ let
+            {
+              # Enable encryption by default,
+              options.forceSSL.default = true;
+              options.enableACME.default = true;
```
I know I have introduced this option, but on a second thought, it is quite unnecessary to force these settings upon user. Wouldn't it be reasonable to drop it - along with forceSSL?

But in that case, I guess: `fastcgi_param HTTPS on;` would have to go in that case, too.

dadada replied:
I guess it would be wise to go with the default that works in most cases. ACME does not always work depending on restrictive DNS CA records or missing internet connection.

The [default](https://nixos.org/nixos/options.html#forcessl) seems to be to disable ACME and not to force SSL. But it should be enabled in the example.

Regarding [ninx fastcgi](https://nginx.org/en/docs/http/ngx_http_fastcgi_module.html), if_not_empty should do the trick.

>  If the directive is specified with if_not_empty (1.1.11) then such a parameter will be passed to the server only if its value is not empty:
>
>    fastcgi_param HTTPS           $https if_not_empty;

1000101 replied:
Sounds about right, let's do it that way.

dadada replied:
Since would be an independent change, so you could open a new MR for this.

1000101 replied:
Sure, could do that, but since you did such a great job in refactoring a few things here and there, I though it would be a nice addition to the PR. But you're right, we can definitely keep it separated, it's just nice to have and not subject of this PR.

dadada commented:
> You mean users.auth.php (i.e. `usersFile` option)? It was actually not part of the derivation so that 1. credentials would not end up in nix store 2. it would be mutable (users could change their passwords).

Ok, I might have misunderstood how `types.path` works. I was under the impression that the path will automatically be copied to the derivation / package, but this does not seem to be the case.
 
> Generally, I was trying to have as much options as I can immutable, with obvious ones mutable (https://github.com/NixOS/nixpkgs/pull/77830/files#r368233914 (resolved conversation) ).

I made some changes so that the mutable files for `usersFile` and `aclFile` will not be automatically created unless specified, which should be identical to how these parameters behaved in your version.

1000101 reviewed (changes requested):
Other than that, works really great, awesome job!!!

1000101 commented:
> > You mean users.auth.php (i.e. `usersFile` option)? It was actually not part of the derivation so that 1. credentials would not end up in nix store 2. it would be mutable (users could change their passwords).
> 
> Ok, I might have misunderstood how `types.path` works. I was under the impression that the path will automatically be copied to the derivation / package, but this does not seem to be the case.
> 
> > Generally, I was trying to have as much options as I can immutable, with obvious ones mutable (https://github.com/NixOS/nixpkgs/pull/77830/files#r368233914 (resolved conversation) ).
> 
> I made some changes so that the mutable files for `usersFile` and `aclFile` will not be automatically created unless specified, which should be identical to how these parameters behaved in your version.

Oh, I actually meant to say that I think your approach is correct (much better than mine, for that matter) as these files stay outside of /nix/store (in datadir) and they don't need to be created by hand. I'm sorry I didn't make myself clear!!! I think you're right the whole time, or am I missing something? :)))

dadada commented:
> Oh, I actually meant to say that I think your approach is correct (much better than mine, for that matter) as these files stay outside of /nix/store (in datadir) and they don't need to be created by hand. I'm sorry I didn't make myself clear!!! I think you're right the whole time, or am I missing something? :)))

Thanks, but there might be two aspects here:

1. private data is not stored in the nix store and second
2. the configuration is valid (the paths for `usersFile` and `aclFile` exist)

1. With your approach `usersFile` and `aclFile` already were outside of /nix/store if specified, so everything should be fine. Otherwise they would be immutable, so no private data would end up in the nix store. Unless you consider the contents of `acl` to be private, hence the added warning in the description.

2. Since `usersFile` and `aclFile` are both types.path, they are checked at configuratio time and if they don't exists, nixos-reconfigure fails, which is good. So no problem here either. The same would happen with the changes I made (still types.path).

I think the best would be to not create them, if `usersFile` or `aclFile` was not specified, but make them both types.str, so building the configuration does not fail, but the files are created if they don't exist at service startup.

1000101 commented:
@GrahamcOfBorg test dokuwiki

dadada commented:
> I think the best would be to not create them, if `usersFile` or `aclFile` was not specified, but make them both types.str, so building the configuration does not fail, but the files are created if they don't exist at service startup.

On the other hand: If the file can't be created or has the wrong permissions, the service will fail at startup, which makes this kind-of dependent on the system and not 100% reproducible. I think the wireguard module does a simmilar thing with types.str and the private key.

1000101 commented:
> > I think the best would be to not create them, if `usersFile` or `aclFile` was not specified, but make them both types.str, so building the configuration does not fail, but the files are created if they don't exist at service startup.
> 
> On the other hand: If the file can't be created or has the wrong permissions, the service will fail at startup, which makes this kind-of dependent on the system and not 100% reproducible. I think the [wireguard module](https://github.com/NixOS/nixpkgs/blob/74c0ce53763218b591fa4aac5899b9dfdb95f90f/nixos/modules/services/networking/wireguard.nix) does a similar thing with `types.str` and the private key.

Sure, but what about leaving default path set to `datadir`, i.e. `/var/lib/dokuwiki/${name}` ant not null? Would that hurt anything? I think it's quite neat that it kind of works right out of the box, but the user can still change it if needed.

dadada commented:
> > > I think the best would be to not create them, if `usersFile` or `aclFile` was not specified, but make them both types.str, so building the configuration does not fail, but the files are created if they don't exist at service startup.
> > 
> > 
> > On the other hand: If the file can't be created or has the wrong permissions, the service will fail at startup, which makes this kind-of dependent on the system and not 100% reproducible. I think the [wireguard module](https://github.com/NixOS/nixpkgs/blob/74c0ce53763218b591fa4aac5899b9dfdb95f90f/nixos/modules/services/networking/wireguard.nix) does a similar thing with `types.str` and the private key.
> 
> Sure, but what about leaving default value of `usersFile` and `aclFile` set to `datadir`, i.e. `/var/lib/dokuwiki/${name}` ant not null (as you had in previous commits)? Would that hurt anything? I think it's quite neat that it kind of works right out of the box, but the user can still change it if needed.

I thought about it, but that would mean creating `aclFile` even if `acl` is used. So changes to `aclFile` would not affect anything. Same if `aclUse` is false, there would still be a `usersFile`.

dadada reviewed (commented):
How does one access `aclUse` inside of the function. The way I tried it does not seem to work. I don't get how the function is called / which arguments are passed to it.

1000101 commented:
> > > > I think the best would be to not create them, if `usersFile` or `aclFile` was not specified, but make them both types.str, so building the configuration does not fail, but the files are created if they don't exist at service startup.
> > > 
> > > 
> > > On the other hand: If the file can't be created or has the wrong permissions, the service will fail at startup, which makes this kind-of dependent on the system and not 100% reproducible. I think the [wireguard module](https://github.com/NixOS/nixpkgs/blob/74c0ce53763218b591fa4aac5899b9dfdb95f90f/nixos/modules/services/networking/wireguard.nix) does a similar thing with `types.str` and the private key.
> > 
> > 
> > Sure, but what about leaving default value of `usersFile` and `aclFile` set to `datadir`, i.e. `/var/lib/dokuwiki/${name}` ant not null (as you had in previous commits)? Would that hurt anything? I think it's quite neat that it kind of works right out of the box, but the user can still change it if needed.
> 
> I thought about it, but that would mean creating `aclFile` even if `acl` is used. So changes to `aclFile` would not affect anything. Same if `aclUse` is false, there would still be a `usersFile`.

Ok, let's leave it be, it's just a minor nitpicking anyway.

dadada committed:
nixos/dokuwiki: change default of aclFile and usersFile

`aclFile` and `usersFile` will be set to a default value if `aclUse` is
specified and aclFile is not overriden by `acl`.

dadada commented:
Now `aclUse = true;` should be enough to create `users.auth.php` and `acl.auth.php` in the default locations.

1000101 commented:
@GrahamcOfBorg test dokuwiki

1000101 commented:
* [x] reviewed the diff and commit messages
* [x] made sure ofBorg build succeeded for all applicable platforms
* [x] ran nix-review without any failures
* [x] ran and tested the service

dadada commented:
Rebased changes onto master and squashed fixup commits.

1000101 commented:
> Rebased changes onto master and squashed fixup commits.

Oh, and you managed to find a way to access config values, great job!

1000101 commented:
@GrahamcOfBorg test dokuwiki

1000101 reviewed (approved):
* [x] reviewed the diff and commit messages
* [x] made sure ofBorg build succeeded for all applicable platforms
* [x] ran nix-review without any failures
* [x] ran and tested the service

1000101 commented:
@mmahut could you please help us check and merge if everything's fine by you?
